aspnetcore/blazor/components/render-modes.md

```diff
@@ -284,6 +312,41 @@ In the following example, the component is interactive throughout the process. T
 
 If using the preceding component locally in a Blazor Web App, place the component in the client project's `Pages` folder. The client project is the solution's project with a name that ends in `.Client`. When the app is running, navigate to `/render-mode-4` in the browser's address bar.
 
+## Client-side services fail to resolve during prerendering
+
+Assuming that prerendering isn't disabled for a component or for the app, a component in the `.Client` project is prerendered on the server. Because the server doesn't have access to registered client-side Blazor services, it isn't possible to inject these services into a component without receiving an error that the service can't be found during prerendering.
+
+For example, consider the following `Home` component in the `.Client` project in a Blazor Web App with [global Interactive WebAssembly or Interactive Auto rendering](#apply-a-render-mode-to-the-entire-app). The component attempts to inject <xref:Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment> to obtain the environment's name.
+
+```razor
+@page "/"
+@inject IWebAssemblyHostEnvironment Environment
+
+<PageTitle>Home</PageTitle>
+
+<h1>Home</h1>
+
+<p>
+    Environment: @Environment.Environment
+</p>
+```
+
+No compile time error occurs, but a runtime error occurs during prerendering:
+
+> :::no-loc text="Cannot provide a value for property 'Environment' on type 'BlazorWebAppSample.Client.Pages.Home'. There is no registered service of type 'Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment'.":::
+
+This error occurs because the component must compile and execute on the server during prerendering, but <xref:Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment> isn't a registered service on the server.
+
+There are a four approaches that you can take to address this scenario. The following are listed from most recommended to least recommended:
```

Comment:
Often a simpler option is to inject `IServiceProvider` instead of the service type itself, and then you can have code like:

```cs
if (Services.GetService<IWebAssemblyHostEnvironment>() is {} env)
{
   // ... do stuff with env
}
```

TBH that seems easier than making a new abstraction.

Reply:
AH! I see ... yes ... ***BUT***, it won't resolve an environment at all during prerendering. If the dev is counting on the value there, this will 💥 break their logic, correct?